Add blog post about missing side effects on selector access bug (#7)
ethereum/solidity-website: src/posts/2023-07-19-missing-side-effects-on-selector-access-bug.md

Any actually affected coding pattern would likely be considered an anti-pattern by users and disregarded
in favor of more sensible alternatives.
For example, in an expression like ``f().g.selector`` the result will always be the same regardless
of what ``f()`` returns and it is more natural to perform the operation directly on the contract type,
i.e., ``C.g.selector``.
Furthermore, expressions like ``f().selector`` remain unaffected.
As a result, we have assigned it a severity level of "low".

## Technical Details

The bug manifested itself in cases where it was possible to determine which function is the result
of an expression at compilation time, which allowed the compiler to calculate the selector without
actually evaluating the expression.
The legacy code generator would take that shortcut to generate more efficient code, on the wrong
assumption that the expression must be side-effect free.

In contrast, the new IR-based code generator is primarily concerned with producing correct code and
can defer the task of making it efficient to the Yul optimizer.
This means that it does not have to rely on such assumptions.

> The whole expression is always visited and evaluated in the unoptimized code.

Author: Isn't the opposite? I.e. the via IR pipeline is not affected by the bug, only the legacy codegen.

```suggestion
The whole expression is always visited and evaluated in the optimized code.
```

It will only be removed by the optimizer if it is determined to actually be free of side-effects.

Due to a related bug in the type checker, the bug would also impact
certain expressions whose value could not be determined at compilation time.

> The compiler internally associates function types with function definitions, and values of the same
> function type pointing at different definitions were implicitly convertible to each other,
> resulting in the associated definition being carried over to the resulting type.

Reviewer: Like I said on the chat, this is not really due to the implicit conversion. That affected cases where the expression is actually evaluated. There the problems is just due to the propagation of the definition. Let's reword it like this:

```suggestion
The compiler internally associates function types with function definitions, and the associated
definition would be carried over to the result type of an expression.
If the expression involved multiple functions, the resulting type would not necessarily point at
the right one.
```

For example, even in ``(condition ? C.f : D.f).selector``, where the value of ``condition`` is only

> known at runtime, the compiler was able to determine the common type for ``C.f`` and ``D.f``,

Author: Oh, but wasn't this disallowed by the restrict mobile types PR?

For example, the code below, if compiled via IR will result in `x` being
assigned `42` and contract `D` being created, as a side-effect of evaluating the function call `h()`.
On the other hand, if compiled using the legacy code generation pipeline it will result in `x`
being `0` and `D` not being deployed.

```solidity
contract D {
    function f() external {}
}

contract C {
    uint256 x;

    function f() public {
        h().f.selector;
    }

    function h() public returns (D) {
        x = 42;
        return new D();
    }
}
```

Here's another example that uses an assignment expression instead of a function call expression.
In the example below, if the legacy code generation pipeline was used to compile the code, the expression `x = true`
would not be evaluated properly, causing the value of `x` to remain unchanged.

```solidity
contract D {
    function f() external{}
}

contract C {
    bool x;

    function f() public {
        ((x = true) ? D.f : D.f).selector;
    }
}
```

However, the fact that this bug remained undiscovered until now indicates that real projects do not seem to be using such

> a pattern, since the expected usage of the `.selector` is in a constant expression.

Author: Right, it was supposed to be a sort of conclusion.

---
layout: post
published: true

> title: 'Bug in Legacy Code Generation When Accessing the .selector Member on Expressions with Side-Effects'

Reviewer 2: ```suggestion
title: 'Bug in Legacy Code Generation When Accessing the .selector Member on Expressions with Side Effects'
```

date: '2023-07-19'
author: Solidity Team
category: Security Alerts
---

On June 26, 2023, a bug in the legacy code generation pipeline of the Solidity compiler was found during
investigation of a security report related to the use of `abi.decode` with a ternary

> expression, that has side-effects, as the type argument.

Author: I agree with the comma suggestion, but why not `side-effects`?

> expression, that has side-effects, as the type argument.

Reviewer: I looked that up and people are using both versions, but apparently the one without hyphen is vastly more common.

> expression, that has side-effects, as the type argument.

Reviewer 2: https://www.merriam-webster.com/dictionary/side%20effect

The legacy code generator was not evaluating complex expressions, like assignments, function calls or conditionals,
whose  `.selector` was being accessed.
This led to the side-effects of such expressions not being executed, and therefore potentially incorrect behavior of
contracts compiled using the legacy pipeline.

> The via IR code generator behaves as expected and evaluates such expressions.

Reviewer 2: ```suggestion
The via-IR code generator behaves as expected and evaluates such expressions.
```

Author: See https://github.com/ethereum/solidity-website/pull/7#discussion_r1265609231 for the reasoning of why we decided to go for `via IR` instead of `via-IR`

Reviewer 2: [Developer Survey](https://soliditylang.org/blog/2023/03/10/solidity-developer-survey-2022-results/): via-IR
[Documentation](https://docs.soliditylang.org/en/v0.8.21/ir-breaking-changes.html): via-ir.

Let's not add yet another confusing term.

The bug has been present since version 0.6.2 of the compiler.
However, to the best of our knowledge, it has never been exploited or reported externally.
Any actually affected coding pattern would likely be considered an anti-pattern by users and disregarded
in favor of more sensible alternatives.

> For example, in an expression like `f().g.selector` the result will always be the same regardless
> of what `f()` returns and it is more natural to perform the operation directly on the contract type,
> i.e., `C.g.selector`.
> Furthermore, expressions like `f().selector` remain unaffected.
> As a result, we have assigned it a severity level of "low".

Reviewer 2: ```suggestion
For example, in an expression like `f().g.selector`, the result will always be the same regardless
of what `f()` returns, and it is more natural to perform the operation directly on the contract type,
i.e., `C.g.selector`.
Furthermore, expressions like `f().selector` remain unaffected.
As a result, we have assigned it a severity level of "low".
```

> On June 26, 2023, a bug in the legacy code generation pipeline of the Solidity compiler was found during
> investigation of a security report related to the use of `abi.decode` with a ternary
> expression that has side effects, as the type argument.
> 
> The legacy code generator was not evaluating complex expressions, like assignments, function calls or conditionals,
> whose  `.selector` was being accessed.
> This led to the side-effects of such expressions not being executed, and therefore potentially incorrect behavior of
> contracts compiled using the legacy pipeline.
> The via IR code generator behaves as expected and evaluates such expressions.
> 
> The bug has been present since version 0.6.2 of the compiler.
> However, to the best of our knowledge, it has never been exploited or reported externally.
> Any actually affected coding pattern would likely be considered an anti-pattern by users and disregarded
> in favor of more sensible alternatives.
> For example, in an expression like `f().g.selector`, the result will always be the same regardless
> of what `f()` returns, and it is more natural to perform the operation directly on the contract type,
> i.e., `C.g.selector`.
> Furthermore, expressions like `f().selector` remain unaffected.
> As a result, we have assigned it a severity level of "low".

Reviewer 2: ```suggestion
On June 26, 2023, a bug in the legacy code generation pipeline of the Solidity compiler was found during
investigation of a security report related to the use of `abi.decode` with a ternary
expression that has side effects, as the type argument.
The bug has been present since version 0.6.2 of the compiler.
However, to the best of our knowledge, it has never been exploited or reported externally.
Any actually affected coding pattern would likely be considered an anti-pattern by users and disregarded
in favor of more sensible alternatives.

The legacy code generator was not evaluating complex expressions, like assignments, function calls or conditionals,
whose  `.selector` was being accessed.
This led to the side-effects of such expressions not being executed, and therefore potentially incorrect behavior of
contracts compiled using the legacy pipeline.
The via IR code generator behaves as expected and evaluates such expressions.

For example, in an expression like `f().g.selector`, the result will always be the same regardless
of what `f()` returns, and it is more natural to perform the operation directly on the contract type,
i.e., `C.g.selector`.
Furthermore, expressions like `f().selector` remain unaffected.
As a result, we have assigned it a severity level of "low".
```

Reviewer 2: @r0qs this way, a user doesn't have to jump between paragraphs to understand wince when it has been present before going into tech details.

Reviewer 2: This way, it also matches @cameel's blogpost structure.

Author: Oh, actually, I think it was better the way it was before, because that explanation of the bug was necessary, in my point of view, to understand why we classify the severity level as low.

Reviewer 2: The explanation is still there, and so it is the "low" classification in the last sentence.

Or are you suggesting to just move to the bottom the following text? If so, I'm cool.
```
The legacy code generator was not evaluating complex expressions, like assignments, function calls or conditionals,
whose  `.selector` was being accessed.
This led to the side-effects of such expressions not being executed, and therefore potentially incorrect behavior of
contracts compiled using the legacy pipeline.
The via IR code generator behaves as expected and evaluates such expressions.
```

Author: This:
```
On June 26, 2023, a bug in the legacy code generation pipeline of the Solidity compiler was found during
investigation of a security report related to the use of `abi.decode` with a ternary
expression that has side effects, as the type argument.

The legacy code generator was not evaluating complex expressions, like assignments, function calls or conditionals,
whose  `.selector` was being accessed.
This led to the side-effects of such expressions not being executed, and therefore potentially incorrect behavior of
contracts compiled using the legacy pipeline.
The via IR code generator behaves as expected and evaluates such expressions.

For example, in an expression like `f().g.selector`, the result will always be the same regardless
of what `f()` returns, and it is more natural to perform the operation directly on the contract type,
i.e., `C.g.selector`.
Furthermore, expressions like `f().selector` remain unaffected.

The bug has been present since version 0.6.2 of the compiler.
However, to the best of our knowledge, it has never been exploited or reported externally.
Any actually affected coding pattern would likely be considered an anti-pattern by users and disregarded
in favor of more sensible alternatives.
As a result, we have assigned it a severity level of "low".

```

## Technical Details

The bug manifested itself in cases where it was possible to determine which function is the result
of an expression at compilation time, which allowed the compiler to calculate the selector without
actually evaluating the expression.

> The legacy code generator would take that shortcut to generate more efficient code, on the wrong

Reviewer 2: ```suggestion
The legacy code generator would take that shortcut to generate more efficient code on the wrong
```

> assumption that the expression must be side-effect free.

Reviewer 2: ```suggestion
assumption that the expression must be side effect free.
```

In contrast, the new IR-based code generator is primarily concerned with producing correct code and
can defer the task of making it efficient to the Yul optimizer.
This means that it does not have to rely on such assumptions.
The whole expression is always visited and evaluated in the unoptimized code.

> It will only be removed by the optimizer if it is determined to actually be free of side-effects.

Reviewer 2: ```suggestion
It will only be removed by the optimizer if it is determined to actually be free of side effects.
```

> ## Technical Details
> 
> The bug manifested itself in cases where it was possible to determine which function is the result
> of an expression at compilation time, which allowed the compiler to calculate the selector without
> actually evaluating the expression.
> The legacy code generator would take that shortcut to generate more efficient code on the wrong
> assumption that the expression must be side effect free.
> 
> In contrast, the new IR-based code generator is primarily concerned with producing correct code and
> can defer the task of making it efficient to the Yul optimizer.
> This means that it does not have to rely on such assumptions.
> The whole expression is always visited and evaluated in the unoptimized code.
> It will only be removed by the optimizer if it is determined to actually be free of side effects.
> 
> Due to a related bug in the type checker, the bug would also impact
> certain expressions whose value could not be determined at compilation time.
> The compiler internally associates function types with function definitions, and the associated
> definition would be carried over to the result type of an expression.
> If the expression involved multiple functions, the resulting type would not necessarily point at
> the right one.
> 
> For example, even in `(condition ? C.f : D.g).selector`, where the value of `condition` is only
> known at runtime, the compiler was able to determine the common type for `C.f` and `D.g`,
> and would use the definition associated with such type to determine the function to take the selector from.
> This bug was also fixed in this release.

Reviewer 2: ```suggestion
## Which Contracts Are Affected?

If you are still utilizing the legacy code generation pipeline and your contract accesses the `.selector`
on expressions that have side effects your code may be vulnerable to this bug.
For example, the code below, if compiled via IR will result in `x` being
assigned `42` and contract `D` being created as a side effect of evaluating the function call `h()`.
On the other hand, if compiled using the legacy code generation pipeline, it will result in `x`
being `0` and `D` not being deployed.

```solidity
contract D {
    function f() external {}
}
contract C {
    uint256 x;
    function f() public {
        h().f.selector;
    }
    function h() public returns (D) {
        x = 42;
        return new D();
    }
}
```

Below, you can find another example that demonstrates the use of an assignment expression instead of a
function call expression.
In the example, if the legacy code generation pipeline was used, the expression `x = true`
would not be evaluated properly, causing the value of `x` to remain unchanged.

```solidity
contract D {
    function g() external {}
}
contract C {
    bool x;
    function f() external {
        ((x = true) ? C.f : D.g).selector;
    }
}
```
```

## Which Contracts Are Affected?

If you are still utilizing the legacy code generation pipeline and your contract accesses the `.selector`

> on expressions that have side-effects your code may be vulnerable to this bug.

Reviewer 2: ```suggestion
on expressions that have side effects your code may be vulnerable to this bug.
```

> For example, the code below, if compiled via IR will result in `x` being

Reviewer 2: ```suggestion
For example, the code below, if compiled with via-IR will result in `x` being
```

Author: See https://github.com/ethereum/solidity-website/pull/7#discussion_r1267791812

> assigned `42` and contract `D` being created, as a side-effect of evaluating the function call `h()`.

Reviewer 2: ```suggestion
assigned `42` and contract `D` being created as a side effect of evaluating the function call `h()`.
```

> On the other hand, if compiled using the legacy code generation pipeline it will result in `x`

Reviewer 2: ```suggestion
On the other hand, if compiled using the legacy code generation pipeline, it will result in `x`
```

> The whole expression is always visited and evaluated in the unoptimized code.
> It will only be removed by the optimizer if it is determined to actually be free of side effects.
> 
> Due to a related bug in the type checker, the bug would also impact
> certain expressions whose value could not be determined at compilation time.
> The compiler internally associates function types with function definitions, and the associated
> definition would be carried over to the result type of an expression.
> If the expression involved multiple functions, the resulting type would not necessarily point at
> the right one.
> 
> For example, even in `(condition ? C.f : D.g).selector`, where the value of `condition` is only
> known at runtime, the compiler was able to determine the common type for `C.f` and `D.g`,
> and would use the definition associated with such type to determine the function to take the selector from.
> This bug was also fixed in this release.
> 
> ## Which Contracts Are Affected?
> 
> If you are still utilizing the legacy code generation pipeline and your contract accesses the `.selector`
> on expressions that have side effects your code may be vulnerable to this bug.
> For example, the code below, if compiled via IR will result in `x` being
> assigned `42` and contract `D` being created as a side effect of evaluating the function call `h()`.
> On the other hand, if compiled using the legacy code generation pipeline, it will result in `x`
> being `0` and `D` not being deployed.
> 
> ```solidity
> contract D {
>     function f() external {}
> }
> 
> contract C {
>     uint256 x;
> 
>     function f() public {
>         h().f.selector;
>     }
> 
>     function h() public returns (D) {
>         x = 42;
>         return new D();
>     }
> }
> ```
> 
> Below, you can find another example that demonstrates the use of an assignment expression instead of a
> function call expression.
> In the example, if the legacy code generation pipeline was used, the expression `x = true`
> would not be evaluated properly, causing the value of `x` to remain unchanged.
> 
> ```solidity
> contract D {
>     function g() external {}
> }
> 
> contract C {
>     bool x;
>     function f() external {
>         ((x = true) ? C.f : D.g).selector;
>     }
> }
> ```

Reviewer 2: ```suggestion
## Which Contracts Are Affected?

If you are still utilizing the legacy code generation pipeline and your contract accesses the `.selector`
on expressions that have side effects your code may be vulnerable to this bug.
For example, the code below, if compiled via IR will result in `x` being
assigned `42` and contract `D` being created as a side effect of evaluating the function call `h()`.
On the other hand, if compiled using the legacy code generation pipeline, it will result in `x`
being `0` and `D` not being deployed.

```solidity
contract D {
    function f() external {}
}

contract C {
    uint256 x;

    function f() public {
        h().f.selector;
    }

    function h() public returns (D) {
        x = 42;
        return new D();
    }
}
```

Below, you can find another example that demonstrates the use of an assignment expression instead of a
function call expression.
In the example, if the legacy code generation pipeline was used, the expression `x = true`
would not be evaluated properly, causing the value of `x` to remain unchanged.

```solidity
contract D {
    function g() external {}
}

contract C {
    bool x;
    function f() external {
        ((x = true) ? C.f : D.g).selector;
    }
}
```
## Technical Details

The bug manifested itself in cases where it was possible to determine which function is the result
of an expression at compilation time, which allowed the compiler to calculate the selector without
actually evaluating the expression.
The legacy code generator would take that shortcut to generate more efficient code on the wrong
assumption that the expression must be side effect free.

In contrast, the new IR-based code generator is primarily concerned with producing correct code and
can defer the task of making it efficient to the Yul optimizer.
This means that it does not have to rely on such assumptions.
The whole expression is always visited and evaluated in the unoptimized code.
It will only be removed by the optimizer if it is determined to actually be free of side effects.

Due to a related bug in the type checker, the bug would also impact
certain expressions whose value could not be determined at compilation time.
The compiler internally associates function types with function definitions, and the associated
definition would be carried over to the result type of an expression.
If the expression involved multiple functions, the resulting type would not necessarily point at
the right one.

For example, even in `(condition ? C.f : D.g).selector`, where the value of `condition` is only
known at runtime, the compiler was able to determine the common type for `C.f` and `D.g`,
and would use the definition associated with such type to determine the function to take the selector from.
This bug was also fixed in this release.
```

> ```

Reviewer 2: Please swap these two sections to match @cameel 's structure.
Users want to know first Which contracts are affected to know they're safe, not the technical details.

In conclusion, the fact that this bug remained undiscovered until now indicates that real projects do not seem to be using such

> a pattern, since the expected usage of the `.selector` is in a constant expression.

Reviewer 2: ```suggestion
a pattern since the expected usage of the `.selector` is in a constant expression.
```

> a pattern since the expected usage of the `.selector` is in a constant expression.

Reviewer: TBH, this sentence still feels like it comes out of nowhere and does not follow logically from the result of the section :)

---
layout: post
published: true
title: 'Bug in legacy code generation when accessing the .selector member on expressions with side-effects'
date: '2023-07-19'
author: Solidity Team
category: Security Alerts
---

On June 26, 2023, a bug in the legacy code generation pipeline of the Solidity compiler was found during
investigation of a security report related to the use of ``abi.decode`` with a ternary
expression as the type argument.

The legacy code generator was not evaluating complex expressions, like assignments, function calls or conditionals,
whose  ``.selector`` was being accessed.
This led to the side-effects of such expressions not being executed, and therefore potentially incorrect behavior of 
contracts compiled using the legacy pipeline.
The via IR code generator behaves as expected and evaluates such expressions.

The bug has been present since the compiler version ``0.6.2``.
However, to the best of our knowledge, it has never been exploited or reported externally.

> It can be considered an anti-pattern, for example, instead of using ``f().g.selector`` it is more advisable to
> perform the operation directly on the contract, i.e., ``C.g.selector``.

Reviewer: ```suggestion
Any actually affected coding pattern would likely be considered an anti-pattern by users and disregarded on favor of more sensible alternatives.
For example, in an expression like ``f().g.selector`` the result will always be the same regardless of what ``f()`` returns and it is more natural to perform the operation directly on the contract type, i.e., ``C.g.selector``.
```

> Furthermore, expressions like ``f().selector`` remain unaffected.

Reviewer: I think this still does not clarify very well what kind of expressions are affected. We should either give some more general "rule" to determine this, or at least list more examples if giving such a rule in a way understandable to the user is too hard.

For example function pointers are not affected either. What about built-ins and various kinds or internal/external/library functions?

As a result, we have assigned it a severity level of "low".

## Which Contracts are Affected?

If you are still utilizing the legacy code generation pipeline and your contract accesses the ``.selector``
on expressions that have side-effects your code may be vulnerable to this bug.

> For example, the code below, if compiled with ``via-ir`` will result in ``x`` being 
> assigned ``42`` and contract ``D`` being created, as a side-effect of evaluating the function call ``h()``.
> On the other hand, if compiled without ``via-ir`` (i.e. using the legacy code generation pipeline) 

Reviewer: I guess you're referring to the CLI option here?
```suggestion
For example, the code below, if compiled with ``--via-ir`` will result in ``x`` being 
assigned ``42`` and contract ``D`` being created, as a side-effect of evaluating the function call ``h()``.
On the other hand, if compiled without ``--via-ir`` (i.e. using the legacy code generation pipeline) 
```

Author: I was referring to the compiler feature, actually. Which you can enable using the CLI or the Standard JSON. Maybe it should be something like `...if you are compiling via IR...`.

Reviewer: Oh. In that case that would be good too. I generally find "via-IR" used as a noun a bit awkward and try to avoid it. Instead I usually say something like "IR-based code generation pipeline" or "IR generator". On the other hand various spellings of "viair" is what we call the options and often use in external communication so it's getting complicated :)

whose  `.selector` was being accessed.
This led to the side-effects of such expressions not being executed, and therefore potentially incorrect behavior of
contracts compiled using the legacy pipeline.

> The via-IR code generator behaves as expected and evaluates such expressions.

Reviewer 2: [Developer Survey](https://soliditylang.org/blog/2023/03/10/solidity-developer-survey-2022-results/): via-IR
[Documentation](https://docs.soliditylang.org/en/v0.8.21/ir-breaking-changes.html): via-ir.

Let's not add yet another confusing term.
